fabric/src/main/java/hardcorequesting/fabric/tileentity/QSDFluidTank.java

```diff
@@ -0,0 +1,89 @@
+package hardcorequesting.fabric.tileentity;
+
+import com.ibm.icu.impl.Pair;
+import dev.architectury.fluid.FluidStack;
+import hardcorequesting.common.quests.data.ItemsTaskData;
+import hardcorequesting.common.quests.task.QuestTask;
+import hardcorequesting.common.quests.task.item.ConsumeItemTask;
+import hardcorequesting.common.quests.task.item.ItemRequirementTask;
+import net.fabricmc.fabric.api.transfer.v1.fluid.FluidConstants;
+import net.fabricmc.fabric.api.transfer.v1.fluid.FluidVariant;
+import net.fabricmc.fabric.api.transfer.v1.storage.StorageView;
+import net.fabricmc.fabric.api.transfer.v1.storage.base.BlankVariantView;
+import net.fabricmc.fabric.api.transfer.v1.storage.base.InsertionOnlyStorage;
+import net.fabricmc.fabric.api.transfer.v1.transaction.TransactionContext;
+import net.fabricmc.fabric.api.transfer.v1.transaction.base.SnapshotParticipant;
+
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+@SuppressWarnings("UnstableApiUsage")
+public class QSDFluidTank extends SnapshotParticipant<QSDFluidTank.FluidQuestData> implements InsertionOnlyStorage<FluidVariant> {
+    private final List<StorageView<FluidVariant>> blankView = List.of(new BlankVariantView<>(FluidVariant.blank(), FluidConstants.BUCKET));
+    private final BarrelBlockEntity blockEntity;
+
+    public QSDFluidTank(BarrelBlockEntity blockEntity) {
+        this.blockEntity = blockEntity;
+    }
+
+    @Override
+    public long insert(FluidVariant resource, long maxAmount, TransactionContext transaction) {
+        QuestTask<?> questTask = blockEntity.getCurrentTask();
+        if (questTask instanceof ConsumeItemTask task) {
+            FluidStack fluidStack = FluidStack.create(resource.getFluid(), maxAmount);
+            updateSnapshots(transaction);
+            task.increaseFluid(fluidStack, blockEntity.getPlayerUUID(), true);
```

Comment:
`increaseFluid()` can call `doCompletionCheck()`, which I believe should only get called when the transaction has committed.